### ISSUE TITLE
Platform: BSD Port: Current Issues

### ISSUE THREAD

**Issue by bedwardly-down** [thunder-engine/thunder#770]
## Description

Just a general list of bugs I'm working on ironing out. None of these will break the core platform support but now that UIKit's been implemented, everything can be built and tested. I hope to get this fully functional by next main release. There's not many official game dev tools that support the BSDs.

## Prerequisites

Current master has the code changes. Anyone that has 64-bit FreeBSD 14.0 or 14.1 (or other 64-bit BSDs) can test this out.

## Steps to Reproduce

1. Install all of the qt5 requirements (installing the meta package should cover this)
2. Install `libinotify, amath, llvm15+, gcc13+, cmake` (all required)
3. Optional installs: 
* `assimp` (may need to rely on a system version for this one instead of the one under thirdparty; will give details below)
* `ccache` (to speed up compile times; if you use this, add `-DCMAKE_COMPILER_LAUNCHER=ccache` to the end of Step 7 below)
4. Clone the latest master: `git clone --recursive https://github.com/thunder-engine/thunder.git`
5. Make an external build directory: `mkdir -pv thunder-build`
6. `cd thunder-build`
7. Clang is the default compiler and what the repos will use for building if this were packaged by the OS maintainers and community, but I'm finding that GCC is picking up some linkage errors that Clang is ignoring. Run: `CC=gcc CXX=g++ cmake ../thunder` to set everything up for building.
8. I'm not sure about the other BSDs but FreeBSD uses gmake for building native software. Run: `gmake` to start building. You can speed up the build by running `gmake -j <number of cores your CPU has>` instead, but for debugging and testing, that can make it hard to see what's happening.

### Expected behavior:

Everything should compile without issues and WorldEditor should start up without crashing or major errors. None of the dependencies are directly problematic for the BSDs and (at least on FreeBSD) most of them have official ports available.

### Actual behavior:

- [x] 1. Engine editor code ignores QTCore and QTWidgets headers (see terminal snippet below). Both are found under `/usr/local/include/qt5` on my system and are the only part of engine that's having problems at the moment.

```
[ 54%] Building CXX object engine/CMakeFiles/engine.dir/src/editor/assetconverter.cpp.o
In file included from /home/brad/git/thunder/engine/src/editor/assetconverter.cpp:1:
/home/brad/git/thunder/engine/includes/editor/assetconverter.h:4:10: fatal error: 'QObject' file not found
#include <QObject>

[ 57%] Building CXX object engine/CMakeFiles/engine.dir/src/editor/asseteditor.cpp.o
In file included from In file included from /home/brad/git/thunder/engine/src/editor/assetconverter.cpp:1:
/home/brad/git/thunder/engine/includes/editor/assetconverter.h:In file included from /home/brad/git/thunder/engine/src/editor/assetmanager.cpp:1:
/home/brad/git/thunder/engine/includes/editor/assetmanager.h:4:10: 4:/home/brad/git/thunder/engine/src/editor/asseteditor.cpp10::fatal error: 1 'QObject' file not foundfatal error: 
'QObject' file not found
#include <QObject>
/home/brad/git/thunder/engine/includes/editor/asseteditor.h:4         ^~~~~~~~~:
10: fatal error: 'QWidget' file not found
#include <QWidget>
```

- [x] 2. `Assimp::StackAllocator:g_maxBytesPerBlock` throws an undefined reference error when linking WorldEditor at the end through `libengine-editor.so` and sometimes `libgraph-editor.so`. One suggestion from the assimp issue tracker for similar issues was to upgrade to the latest version. I plan to test it out since the version used in the engine is just one below but also expect breakage when doing such.

- [X] 3. WorldEditor crashes due to AddressOutOfBounds segfault after a full successful build with GCC13.

- [ ] 4. Clang fully builds and runs both the WorldEditor and Builder applications but WorldEditor crashes after creating a New Project with this error:

```
 The required settings was not specified: Company Name, Project Version, First Map Please specify them in the Project Settings.
 [PluginManager] Can't load plugin: /home/brad/git/thunder-build/worldeditor/uikit-editor.so
 [ FileIO ] Can't open file index
fish: Job 1, './worldeditor/WorldEditor' terminated by signal SIGSEGV (Address boundary error)
```

- [ ] 5. Uikit-editor.so plugin doesn't build at all. See here for commentary on it: https://github.com/thunder-engine/thunder/issues/770#issuecomment-2199283368

- [x] 6. $ORIGIN linkage needs to be updated to work with the platform. There was a change awhile back to it (wish I had saved my notes on it better) that required a few lines changed to get binaries to look in the correct locations for library files.

- [X] 7. Optional: build local QT5 release to link against for SDK purposes. - Not worried about this. Most BSD devs will handle that themselves.

- [X] 8. Assimp StackAllocator FBXImporter issue is a priority to get figured out. I'm pretty certain it's because the library is being built as a static one and removing the `-fPIC` compile flag throws this error across multiple files:

```
ld: error: relocation R_X86_64_64 cannot be used against local symbol; recompile with -fPIC
>>> defined in ../thirdparty/assimp/libassimp.a(unzip.c.o)
>>> referenced by unzip.c
>>>               unzip.c.o:(unzOpenCurrentFile3) in archive ../thirdparty/assimp/libassimp.a
```
 
### Reproducibility:

All easily reproducible on Clang17 and GCC13 (while Clang 15 and 16 miss some bugs)

## Versions

![Version](https://github.com/thunder-engine/thunder/assets/59133880/a53d61fc-ae28-4f74-8190-757c29ff250c)

**Comment by bedwardly-down** [comment]
Right now, doing a reinstall of all of Qt5 and then rebuilding. It's possible that while I was at work, there was a version update that broke the build since that issue didn't occur before work during my tests.

**Comment by bedwardly-down** [comment]
Update: I have figured out the Assimp issue. Despite this being the error that shows

```
[ 96%] Linking CXX executable WorldEditor
/usr/local/bin/ld: ../engine/libengine-editor.so: undefined reference to `Assimp::StackAllocator::g_maxBytesPerBlock'
collect2: error: ld returned 1 exit status
```
it's actually an issue with the FBX Importer. By disabling it only for FreeBSD, the error completely goes away. Other importers use StackAllocator without problems. It's just that one particular one that has a linking issue. It doesn't seem to be fixed in 5.4.1 (the new release), so I may open an issue report upstream.

I have everything fully building now with GCC13. About to test with Clang.

**Comment by bedwardly-down** [comment]
To solve the Address error, I'm currently working my way back a bit on dependencies to see which ones I can remove the -fPIC compiler flag from and still get a full build. That's where I'm guessing the address issues lie.

**Comment by eprikazchikov** [comment]
Hello,
Regarding Issue 4. I have received reports regarding crash on creation of new project. But was unable to reproduce it on my end. Usually it creates a project in any way and restart application solves the problem. But in your case I can see the UIKit was not properly loaded. This is bad. could you please show me the export table for this lib?

**Comment by bedwardly-down** [comment]
> Hello, Regarding Issue 4. I have received reports regarding crash on creation of new project. But was unable to reproduce it on my end. Usually it creates a project in any way and restart application solves the problem. But in your case I can see the UIKit was not properly loaded. This is bad. could you please show me the export table for this lib?

It's not building that plugin library at all. Linking the libuikit-editor.so file in the directory doesn't solve the issue either (which is probably related to $ORIGIN linkage differences with the platform). I need to dive into some of my older code for that but I did figure out awhile back how to get around it. That's going to be a new issue.

Give me a sec on the export table. It may not be relevant due to the plugin not building at all, but I can grab one for WorldEditor and libuikit-editor.so while I get more awake.

**Comment by bedwardly-down** [comment]
> Update: I have figured out the Assimp issue. Despite this being the error that shows
> 
> ```
> [ 96%] Linking CXX executable WorldEditor
> /usr/local/bin/ld: ../engine/libengine-editor.so: undefined reference to `Assimp::StackAllocator::g_maxBytesPerBlock'
> collect2: error: ld returned 1 exit status
> ```
> 
> it's actually an issue with the FBX Importer. By disabling it only for FreeBSD, the error completely goes away. Other importers use StackAllocator without problems. It's just that one particular one that has a linking issue. It doesn't seem to be fixed in 5.4.1 (the new release), so I may open an bug report upstream.
> 
> I have everything fully building now with GCC13. About to test with Clang.

Building as a shared library with the -editor.so suffix does solve the compiler errors but doesn't solve the linkage issue. It's probably not a viable solution.

**Comment by bedwardly-down** [comment]
Full stack trace from GDB at crash:

```
Thread 1 received signal SIGSEGV, Segmentation fault.
Address not mapped to object.
0x0000000824420e4b in PluginManager::createRenderer() const ()
   from /home/brad/git/thunder-build/engine/libengine-editor.so
```

**Comment by bedwardly-down** [comment]
With this [PR](https://github.com/thunder-engine/thunder/pull/771), all of the port specific bugs here have been merged into master. Number 4 and 5 are most likely engine specific and not from the port work since neither of them were directly altered by code I've submitted. The PluginManager::createRenderer() crash could possibly be related to [this](https://github.com/thunder-engine/thunder/issues/739) issue. Finding out more before continuing to look into it.
